pkg/service/operation/bolt.go

```diff
@@ -0,0 +1,135 @@
+package operation
+
+import (
+	"strings"
+
+	"github.com/goccy/go-json"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/tbd54566975/ssi-service/internal/util"
+	opstorage "github.com/tbd54566975/ssi-service/pkg/service/operation/storage"
+	"github.com/tbd54566975/ssi-service/pkg/service/operation/storage/namespace"
+	"github.com/tbd54566975/ssi-service/pkg/service/operation/submission"
+	"github.com/tbd54566975/ssi-service/pkg/storage"
+	"go.einride.tech/aip/filtering"
+)
+
+const (
+	cancelledReason = "operation cancelled"
+)
+
+type BoltOperationStorage struct {
+	db *storage.BoltDB
+}
+
+func (b BoltOperationStorage) CancelOperation(id string) (opstorage.StoredOperation, error) {
+	if strings.HasPrefix(id, submission.ParentResource) {
+		_, opData, err := b.db.UpdateValueAndOperation(
+			submission.Namespace, submission.ResourceID(id), storage.NewUpdater(map[string]any{
+				"status": submission.StatusCancelled,
+				"reason": cancelledReason,
+			}),
+			namespace.FromID(id), id, submission.OperationUpdater{
+				UpdaterWithMap: storage.NewUpdater(map[string]any{
+					"done": true,
+				}),
+			})
+		if err != nil {
+			return opstorage.StoredOperation{}, errors.Wrap(err, "updating value and op")
+		}
+		var op opstorage.StoredOperation
+		if err = json.Unmarshal(opData, &op); err != nil {
+			return op, errors.Wrap(err, "unmarshalling data")
+		}
+		return op, nil
+	}
+	return opstorage.StoredOperation{}, errors.New("unrecognized id structure")
+}
+
+func (b BoltOperationStorage) StoreOperation(op opstorage.StoredOperation) error {
+	id := op.ID
+	if id == "" {
+		return util.LoggingNewError("ID is required for storing operations")
+	}
+	jsonBytes, err := json.Marshal(op)
+	if err != nil {
+		return util.LoggingErrorMsgf(err, "marshalling operation with id: %s", id)
+	}
+	return b.db.Write(namespace.FromID(id), id, jsonBytes)
```

Comment:
is it worth pulling this off as a separate value to be able to wrap the error with a log in case it fails?